zaps/0002-encrypt-push-notifications.md

```diff
@@ -0,0 +1,251 @@
+    user: ForeignKey[UserProfile]
+    encrypted_push_registration: Ciphertext  # A PushRegistration, encrypted with bouncer_key.
+    push_account_id: int  # 64-bit random integer generated by client; need not be unique.
+    bouncer_device_id: RandomID[32] | None  # Generated by bouncer as its unique ID for this record. Null until bouncer has confirmed the registration.
+    push_public_key: PublicKey  # Public part of the asymmetric key pair generated by the client.
+    last_accessed: datetime  # The last time it was accessed to send notification.
+  ```
+- On the push bouncer, add a new table (parallel to `RemotePushDeviceToken`):
+  ```
+  class zilencer.models.RemotePushDevice:
+    device_id: int  # 64-bit integer primary key assigned by bouncer; not random.
+    realm: ForeignKey[RemoteRealm]
+    token_kind: "apns" | "fcm"
+    token: str  # Provided by client.
+    push_account_id: int  # 64-bit random integer generated by client; need not be unique.
+    expired_time: datetime | None
+  ```
+
+- Make sure to correctly maintain the foreign key from `PushDevice`​ to
+  `UserProfile` during export/import. If not done carefully, it can leak
+  notifications to the wrong recipients.
+
+- Zulip Cloud doesn't use a third-party bouncer. To keep things conceptually
+  similar to the self-hosted case with the fewest different scenarios to think
+  through (reducing the risk of bugs), add a new table on the server (parallel
+  to `RemotePushDevice` added in the previous step).
+
+- We probably also need a `zerver.models.push_notifications.PushDevice`
+  for Zulip Cloud use with no third-party bouncer? Or we could make Cloud use a bouncer.
+
+- Additionally, the push bouncer has an asymmetric key `bouncer_key`,
+  of which the public part is baked into clients.
+
+## Operation
+
+### To register a new push device for a Zulip account
+
+1. The client checks the [`get-server-settings`](https://zulip.com/api/get-server-settings)
+   response to see whether the server advertises the `encrypted_push_notifications`
+   capability (or has a new enough feature level, details TBD).
+
+   If it does not, it uses the legacy push notifications registration flow.
+   (Note: We’ll likely want settings to require E2E encryption; that can come
+   after the feature is complete).
+
+1. The client generates a random 64-bit integer `push_account_id`.
+   It builds a JSON object `push_registration` of type `PushRegistration`,
+   and encrypts it to `bouncer_key` as `encrypted_push_registration`.
+
+   It also generates an asymmetric key pair (`push_public_key` and
+   `push_private key`).
+
+1. The client makes an API call to the server (authenticated with the user’s
+   API key) with arguments `push_account_id`, `encrypted_push_registration`,
+   and `push_public_key`.
+
+   The server creates a `PushDevice` instance. It replies to the client with a
+   success response, and asynchronously initiates the next step.
```

Comment:
While the 500s from the bouncer service do not require the client to retry, what happens if it's the server that's getting 500s? Perhaps the client can keep using the same `push_account_id` until it is successful.

Reply:
Possibly clients should treat 500s as a transient error that they should retry the same as a network error?

Reply:
Yeah, the client will keep retrying with the same `push_account_id` until it receives a success response.

I'll post an edit to steps 2 and 3 to help clarify this.

Reply:
(That edit: https://github.com/zulip/zulip-architecture/pull/2#discussion_r2057020659 .)